[Tee Yi Teng] iP

Puakii opened the pull request:
### DukePro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,
1. download it from [here](https://nus-cs2103-ay2223s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉 

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Puakii committed:
Level 1

Puakii committed:
Level-2

Puakii committed:
Level 3. Mark as Done

Puakii committed:
Structure code into better OOP format

Puakii committed:
Level 4. ToDos, Events, Deadlines

Puakii committed:
Level 4. ToDos, Events, Deadlines

Puakii committed:
A-TextUiTesting

Puakii committed:
Level 5. Handle Errors

Puakii committed:
Level 6. Delete

Puakii committed:
A-Enums

Puakii committed:
Updated EXPECTED.TXT

Puakii committed:
Add saving and loading of tasks into hard disk

Puakii committed:
Update Date and Time to use LocalDate

Puakii committed:
Merge branch 'branch-Level-7'

Puakii committed:
Merge branch 'branch-Level-8'

Puakii committed:
Add more OOP

Puakii committed:
Organise code into packages

Puakii committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

Puakii committed:
Edit Gradle configuration

Puakii committed:
Add JUnit tests

Puakii committed:
Add Javadocs for Duke class

Puakii committed:
Add Javadocs for all classes

Puakii committed:
Modify code to comply with coding standard

Puakii committed:
Add ability to find tasks by searching for a keyword.

Puakii committed:
Merge branch 'branch-A-JavaDoc'

Puakii committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/command/Command.java
#	src/main/java/duke/command/EndCommand.java
#	src/main/java/duke/command/ListCommand.java
#	src/main/java/duke/task/Deadline.java

Puakii committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/component/TaskList.java
#	src/main/java/duke/task/Task.java

Puakii committed:
Use CheckStyle

Puakii committed:
Merge branch 'branch-A-CheckStyle'

Puakii committed:
Add GUI to Duke

Puakii committed:
Use Varargs in findTasks

Puakii committed:
Merge branch 'branch-Level-10'

Puakii committed:
Merge branch 'branch-A-Varargs'

ketamethane commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,133 @@
+        TaskType taskType = TaskType.parse(taskDetails[0]);
+        Task currTask;
+        switch (taskType) {
+        case DEADLINE:
+            if (taskDetails.length < 4) {
+                throw new DukeException("Invalid task description!");
+            }
+            currTask = new Deadline(taskDetails[2], taskDetails[3]);
+            break;
+        case EVENT:
+            if (taskDetails.length < 4) {
+                throw new DukeException("Invalid task description!");
+            }
+            currTask = new Event(taskDetails[2], taskDetails[3]);
+            break;
+        case TODO:
+            if (taskDetails.length < 3) {
+                throw new DukeException("Invalid task description!");
+            }
+            currTask = new ToDo(taskDetails[2]);
+            break;
```
Good use of break in the cases! It's a common mistake to leave it out, which may lead to unintended side effects.

marcuslowhuiyu replied:
I agree with him here, forgetting to break after each case has led to some unforeseen errors in my own code as well

ketamethane commented on src/main/java/duke/component/TaskList.java:
```diff
@@ -0,0 +1,148 @@
+    }
+
+    /**
+     * Returns a String representation of the TaskList for storage.
```
Should there be a line break between Lines 49 and 50?

ketamethane reviewed (commented):
I was tasked to review your PR coding standard. Overall, I am left impressed as there was hardly any mistakes spotted at all. Even the indentation is all correct.

marcuslowhuiyu commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,133 @@
+        String[] taskDetails = task.split(" \\| ");
+        TaskType taskType = TaskType.parse(taskDetails[0]);
+        Task currTask;
+        switch (taskType) {
```
I like your use of the switch case statements

marcuslowhuiyu replied:
I like your use of the switch case statements

marcuslowhuiyu commented on src/main/java/duke/task/TaskType.java:
```diff
@@ -0,0 +1,40 @@
+ */
+public enum TaskType {
+
+    DEADLINE("D"),
```
I'm not 100% sure on this, but I think the coding standard mention that enums should be in pascal case, not capital case

marcuslowhuiyu commented on src/main/java/duke/component/Parser.java:
```diff
@@ -0,0 +1,76 @@
+import duke.command.MarkCommand;
+import duke.command.ToDoCommand;
+import duke.command.UnmarkCommand;
+import duke.exception.DukeException;
```
I think it's a good practice to import everything one by one

Puakii committed:
Add assertions

Parser and Task classes have methods that require parameters given to
be valid for successful execution.

These methods either fails or does not make sense if the given
parameters are null.

Assertions are added to them to define assumptions about the program
state to ensure that parameters are confirmed to be valid.

Puakii committed:
Improve Code Quality

Code currently contains unnecessary usage of keyword 'this' as well as
long methods.

Remove unnecessary 'this' keywords as well as structure long methods
in sections with a line space.

Doing so improves code quality and readability.

Puakii committed:
Improve Code Quality

Code currently contains unnecessary usage of keyword 'this' as well as
long methods.

Remove unnecessary 'this' keywords as well as structure long methods
in sections with a line space.

Doing so improves code quality and readability.

Puakii committed:
Add ability to input multiple commands

User can only give command one at a time which can be inefficient.

Add the ability to type multiple commands in a line, splitting each
command using '@'.

This increases the efficiency of user command.

Puakii committed:
Merge pull request #1 from Puakii/branch-A-Assertions

Branch A Assertions

Puakii committed:
Merge branch 'master' of https://github.com/Puakii/ip

* 'master' of https://github.com/Puakii/ip:
  Add assertions

Puakii committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/command/ListCommand.java

Puakii committed:
Merge pull request #2 from Puakii/branch-A-CodeQuality

Branch A Code Quality

Puakii committed:
Merge branch 'master' into branch-C-MassOps

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/command/DeadlineCommand.java
#	src/main/java/duke/command/DeleteCommand.java
#	src/main/java/duke/command/EventCommand.java
#	src/main/java/duke/command/MarkCommand.java
#	src/main/java/duke/command/UnmarkCommand.java
#	src/main/java/duke/component/Parser.java

Puakii committed:
Merge pull request #3 from Puakii/branch-C-MassOps

Add ability to input multiple commands

Puakii committed:
Set up CI

Puakii committed:
Add user guide

Puakii committed:
Update test cases
